### PR TITLE
Consume VS Code built-in extensions in Che-Theia

### PR DESCRIPTION
**Pull request by vzhukovskii**
### What does this PR do?
This changes proposal include in `che-theia` image default VS Code built-in extensions that provide by theia build.

To check this PR, you can use following devfile:

[Try Che-Theia editor only](https://che.openshift.io/f/?url=https://github.com/vzhukovskii/devfiles/raw/master/workspace.yaml)

When workspace will start, you'll be able to open `View` - `Open View...` - `Plugins` to ensure that all VS Code built-in extensions are included into assembly.

Signed-off-by: Vladyslav Zhukovskyi <vzhukovs@redhat.com>

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/14759

#### Release Notes
Consume VS Code built-in extensions in Che-Theia

#### Docs PR
N/A - internal enhancement

### CODE REVIEW

**Comment by vzhukovskii** [comment]
crw-ci-test

File: generator/src/foreach_yarn [hunk 1]
```diff
@@ -45,3 +49,14 @@ function mkdirSyncRecursive(dir) {
         curPath.length > 0 && !fs.existsSync(curPath) ? fs.mkdirSync(curPath) : null;
     }
 };
+
+function copyFolderSync(from, to) {
```
Comment (benoitf) [inline generator/src/foreach_yarn]: hello, is it fast to copy all the plugins ?

Comment (vzhukovskii) [reply generator/src/foreach_yarn]: You mean, in comparison with calling system `cp -r` ?

Comment (benoitf) [reply generator/src/foreach_yarn]: yes sorry I didn't write much comment

yes comparing this nodejs copy vs system cp

Comment (vzhukovskii) [reply generator/src/foreach_yarn]: Haven't checked yet. I can try to compare these two approach.

**Comment by azatsarynnyy** [comment]
If I understand everything right, we should not just copy the [set of all the VS Code extensions](https://github.com/eclipse-theia/theia/blob/86704a9ca6106ca5c9c34f7b42c6066b82cb1d82/package.json#L79) that Theia includes.
But we need to have our own set of VS Code extensions that we want to include in Che Theia.
@benoitf WDYT?

If that's correct, we can just reuse the mechanism provided by Theia:
- in [Che Theia assembly](https://github.com/eclipse/che-theia/blob/master/generator/src/templates/assembly-package.mst) add `theiaPlugins` field like in Theia's [package.json](https://github.com/eclipse-theia/theia/blob/86704a9ca6106ca5c9c34f7b42c6066b82cb1d82/package.json#L79)
- use [`download:plugins` script](https://github.com/eclipse-theia/theia/blob/86704a9ca6106ca5c9c34f7b42c6066b82cb1d82/package.json#L71) to fetch all the specified extensions

**Comment by benoitf** [comment]
@azatsarynnyy maybe that yes if we're depending on a custom set of extensions it'll be easier

It's mainly because we decided to move some extensions to other containers
about your proposal we may check if it'll download the extensions everytime or if there is cache on downloads as well.

Also probably we need to remove https://github.com/eclipse/che-theia/blob/master/dockerfiles/theia/Dockerfile#L99 and use it from https://github.com/theia-ide/vscode-builtin-extensions/releases as well (through package.json `theiaPlugins`

**Comment by vzhukovskii** [comment]
@benoitf we need to take into account, that if we use `download:plugins` command in our package.json, we should also rely on specific url for the every extension. Which may be updated in upstream. I think, we should have somewhere the list of name extensions, which we agreed to use with che-theia and copy them from theia package.

**Comment by vzhukovskii** [comment]
crw-ci-test

**Comment by vzhukovskii** [comment]
@benoitf I compared two ways of copying the directories:

called command to measure the execution time and copied 46 extensions:
`time ./foreach_yarn`

nodejs copy with function:
```
function copyFolderSync(from, to) {
    fs.mkdirSync(to);
    fs.readdirSync(from).forEach(element => {
        if (fs.lstatSync(path.join(from, element)).isFile()) {
            fs.copyFileSync(path.join(from, element), path.join(to, element));
        } else {
            copyFolderSync(path.join(from, element), path.join(to, element));
        }
    });
}
```
./foreach_yarn  0.12s user 0.17s system 55% cpu 0.524 total
./foreach_yarn  0.13s user 0.18s system 79% cpu 0.389 total

system copy with the function:
```
        const r = spawnSync('cp', ['-R', `${pluginFolderPath}`, `${targetDir}/${pluginFolderName}`], {
            cwd: undefined,
            detached: true,
            stdio: "inherit"
        });
        if (r.status === 1) {
            console.log('this one');
            process.exit(1);
        }
```
./foreach_yarn  0.14s user 0.32s system 81% cpu 0.555 total
./foreach_yarn  0.13s user 0.28s system 84% cpu 0.492 total

So, copying using nodejs a bit faster, as it doesn't creates additional process.

File: generator/src/foreach_yarn (hunk 1 above)
Comment (vzhukovskii) [reply generator/src/foreach_yarn]: Comparison provided in: https://github.com/eclipse/che-theia/pull/618#issuecomment-584079385

File: generator/src/templates/theiaPlugins.json [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    "vscode-builtin-shaderlab": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/shaderlab-1.39.1-prel.vsix",
+    "vscode-builtin-shellscript": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/shellscript-1.39.1-prel.vsix",
+    "vscode-builtin-sql": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/sql-1.39.1-prel.vsix",
+    "vscode-builtin-swift": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/swift-1.39.1-prel.vsix",
+    "vscode-builtin-typescript": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/typescript-1.39.1-prel.vsix",
+    "vscode-builtin-vb": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/vb-1.39.1-prel.vsix",
+    "vscode-builtin-xml": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/xml-1.39.1-prel.vsix",
+    "vscode-builtin-yaml": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/yaml-1.39.1-prel.vsix"
+}
```
Comment (benoitf) [inline generator/src/templates/theiaPlugins.json]: minor: missing EOL

Comment (vzhukovskii) [reply generator/src/templates/theiaPlugins.json]: Added an extra line

**Comment by vzhukovskii** [comment]
crw-ci-test

**Comment by benoitf** [comment]
Hello, FYI when using the image built by CI for this PR there is no change on plug-ins ? (image maxura/che-theia:618 )

in your image I think that there is still an issue as textmate grammar extension is still there while it shouldn't as now grammars are brought by VS Code extensions

**Comment by azatsarynnyy** [comment]
I'm testing it with the Factory link provided by che-bot.

I can see the VSCode extensions are picked-up by Che Theia :+1: 
image

**Comment by azatsarynnyy** [comment]
I've noticed that the time of loading Che Theia is significantly increased, up to several seconds.
Is it because of VS Code extensions are unpacking at runtime?

File: generator/src/templates/theiaPlugins.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (azatsarynnyy) [inline generator/src/templates/theiaPlugins.json]: I think we can also get vscode-git https://github.com/che-incubator/vscode-git/releases through that mechanism.

And remove fetching it during build: https://github.com/eclipse/che-theia/blob/1d2ae629da37a96325bfafc3e1e6f6d01b455bd7/dockerfiles/theia/build.sh#L30-L34
Also, copying: https://github.com/eclipse/che-theia/blob/1d2ae629da37a96325bfafc3e1e6f6d01b455bd7/dockerfiles/theia/Dockerfile#L99

**Comment by azatsarynnyy** [comment]
Since the grammars will be provided by the VS Code builtins, we can get rid of `textmate-grammars` Theia extension:
https://github.com/eclipse/che-theia/blob/1d2ae629da37a96325bfafc3e1e6f6d01b455bd7/generator/src/templates/assembly-package.mst#L45

**Comment by vzhukovskii** [comment]
> Is it because of VS Code extensions are unpacking at runtime?

It's related with the bootstrap flow. Each plugins loads sequentially. And because of huge amount of VSCode builtin extensions bootstrap time might be affected.

File: generator/src/templates/theiaPlugins.json (hunk 3 above)
Comment (vzhukovskii) [reply generator/src/templates/theiaPlugins.json]: Will check this.

**Comment by vzhukovskii** [comment]
@benoitf,

> in your image I think that there is still an issue as textmate grammar extension is still there while it shouldn't as now grammars are brought by VS Code extensions

I'll try to remove it from the assembly and recheck the result

File: generator/src/templates/theiaPlugins.json (hunk 3 above)
Comment (vzhukovskii) [reply generator/src/templates/theiaPlugins.json]: Replaced with provided VSCode builtin extension. It works.

File: generator/src/templates/theiaPlugins.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    "vscode-builtin-ruby": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/ruby-1.39.1-prel.vsix",
+    "vscode-builtin-rust": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/rust-1.39.1-prel.vsix",
+    "vscode-builtin-scss": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/scss-1.39.1-prel.vsix",
+    "vscode-builtin-shaderlab": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/shaderlab-1.39.1-prel.vsix",
+    "vscode-builtin-shellscript": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/shellscript-1.39.1-prel.vsix",
+    "vscode-builtin-sql": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/sql-1.39.1-prel.vsix",
+    "vscode-builtin-swift": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/swift-1.39.1-prel.vsix",
+    "vscode-builtin-typescript": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/typescript-1.39.1-prel.vsix",
```
Comment (azatsarynnyy) [inline generator/src/templates/theiaPlugins.json]: I guess typescript extension can be removed as we're providing Che Plugin instead

Comment (vzhukovskii) [reply generator/src/templates/theiaPlugins.json]: Rmoved

Comment (azatsarynnyy) [reply generator/src/templates/theiaPlugins.json]: I've tested the updated image and there's no syntax highlighting for TS.

As I understand, there're two builtin extensions in VS Code: [typescript-basics](https://github.com/microsoft/vscode/tree/master/extensions/typescript-basics) and [typescript-language-features](https://github.com/microsoft/vscode/tree/master/extensions/typescript-language-features).
What we provide with TS Che Plugin it's `typescript-language-features`. And `typescript-basics` must be bundled with Che Theia.
So, let's revert the removed one - `vscode-builtin-typescript`. Sorry for the confusion.

**Comment by vzhukovskii** [comment]
crw-ci-test

**Comment by vzhukovskii** [comment]
crw-ci-test

**Comment by vzhukovskii** [comment]
crw-ci-test

**Comment by vzhukovskii** [comment]
crw-ci-test

**Comment by RomanNikitenko** [comment]
@vzhukovskii 
I see that you are restarting the tests, but maybe at the moment it's expected that the tests failed
https://github.com/eclipse/che/issues/16017

**Comment by vzhukovskii** [comment]
@RomanNikitenko the latest run was unstable, error message found in logs:
```
17:01:36 @eclipse-che/theia-generator:   ● Test Extensions › use default extensions
17:01:36 @eclipse-che/theia-generator:     Unable to execute the command git clone -b master --single-branch --depth 1 https://github.com/eclipse/che-theia che-theia: Error: Command failed: git clone -b master --single-branch --depth 1 https://github.com/eclipse/che-theia che-theia
17:01:36 @eclipse-che/theia-generator:     fatal: Not a git repository: '/tmp/tmpExtensions13845Z4NR05zaBxQz/che-theia/che-theia/.git'
17:01:36 @eclipse-che/theia-generator:       at cp.exec (src/command.ts:939:20)
```

**Comment by vzhukovskii** [comment]
crw-ci-test

**Pull request merged**